---
title: "Analysis of single-cell RNA-seq data using fastglmpca"
author: Eric Weine & Peter Carbonetto
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_float: no
    highlight: textmate
    theme: readable
vignette: >
  %\VignetteIndexEntry{Analysis of single-cell RNA-seq data using fastglmpca}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

The aim of this vignette is to introduce the basic workflow for
fitting a GLM-PCA model (see [Townes *et al* (2019)][townes-2019] or [Collins *et al* (2001)][collins-2001] for more details) with fastglmpca.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

We begin our analysis by loading the packages. Then we set the seed so
that the results can be reproduced.

```{r load-pkgs, message=FALSE, warning=FALSE}
library(Matrix)
library(fastglmpca)
library(ggplot2)
library(cowplot)
library(Seurat)
set.seed(1)
```

The example data set
--------------------

We will illustrate the concepts using a single-cell RNA-seq data set
from [Zheng *et al* (2017)][zheng-2017]. These data are reference
transcriptome profiles from 10 bead-enriched subpopulations of
peripheral blood mononuclear cells (PBMCs). The original data set is
much larger---for this introduction, we have taken a small subset of
approximately 3,700 cells.

The data we will analyze are unique molecular identifier (UMI)
counts. These data are stored as an $n \times m$ sparse matrix, where
$n$ is the number of genes and $m$ is the number of cells:

```{r load-data}
data(pbmc_facs)
counts <- pbmc_facs$counts
dim(counts)
```

The UMI counts are "sparse"---that is, most of the counts are
zero. Indeed, over 95% of the UMI counts are zero:

```{r}
mean(counts > 0)
```

In order to facilitate faster model fitting, we sample down the genes to the top 3000 "most variable" genes using R package [Seurat][seurat-pkg]. In general, we recommend fitting models on all available genes, but for the sake of speed in demonstrating the use of our package, we downsample.

```{r}
seurat_obj <- CreateSeuratObject(counts = counts)
seurat_obj <- FindVariableFeatures(seurat_obj, nfeatures = 3000)
top_variable_genes <- VariableFeatures(seurat_obj)

counts_samp <- counts[top_variable_genes,]
```

Initializating and fitting a GLM-PCA model
-------------------

Since no pre-processing of UMI counts is needed (e.g. log-transformation), we begin by pre-specifying some "hyper-parameters" of our model. In particular, we must define the rank "K" of our low-rank approximation to the matrix of UMI counts. We provide a convenient function, `init_glmpca_pois`, for doing this.

```{r}
fit0 <- init_glmpca_pois(counts_samp, K = 2)
```

While this function has a simple interface, it does quite a bit of work behind the scenes, and is very flexible. First, by default, `init_glmpca_pois` initializes a gene (row) specific intercept, and a fixed column (cell) specific size-factor (see [Townes *et al* (2019)][townes-2019] for more details). This is meant to imitate the default behavior of R package [glmpca][glmpca-pkg]. `init_glmpca_pois` also provides the functionality to fit row and / or column specific covariates (e.g. batch), though we do not demonstrate the use of these features here.

Once we have initialized our fit, we can carry out model fitting. We do this with `fit_glmpca_pois`, as follows.

```{r}
fit <- fit_glmpca_pois(counts_samp, fit0 = fit0, max_iter = 1000, verbose = FALSE)
```

The output of `fit_glmpca_pois` is similar to most other interfaces that perform PCA. Most users will be interested in `fit$U` and `fit$V`, the orthonormal matrices containing the latent representation of the data. As a basic example, we can extract the loadings for each cell in the first and second directions of variation as follows:

```{r}
pc_df <- data.frame(
  celltype = pbmc_facs$samples$celltype,
  PC1 = fit$V[,1],
  PC2 = fit$V[,2]
)
```

Then, we can visualize these loadings, coloring each point by celltype. 

```{r}
ggplot(pc_df,aes(x = PC1,y = PC2,color = celltype)) +
  geom_point(size = 1) +
  theme_cowplot(font_size = 10)
```

Session info
------------

This is the version of R and the packages that were used to generate
these results.

```{r session-info}
sessionInfo()
```

[zheng-2017]: https://doi.org/10.1038/ncomms14049
[townes-2019]: https://doi.org/10.1186/s13059-019-1861-6
[collins-2001]: https://proceedings.neurips.cc/paper_files/paper/2001/file/f410588e48dc83f2822a880a68f78923-Paper.pdf
[glmpca-pkg]:https://cran.r-project.org/web/packages/glmpca/index.html
[seurat-pkg]:https://cran.r-project.org/web/packages/Seurat/index.html
